### DISCUSSION TITLE
A non immortal PR is reopen multiple times because the first time it was (wrongly) accepted then manually reverted

### DISCUSSION THREAD

**Discussion by jgraglia** [renovatebot/renovate#32564] [Request Help]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

Bitbucket Cloud, Renovate v.39.17.0 

### Please tell us more about your question or problem

Hi,
I'm facing a strange situation where Renovate still recreate & open PR ignoring my decline.
I think the origin of the problem was that the first PR to update the lib (PR  name is "Update dependency com.github.jknack:handlebars to v4.3.1" (branch: renovate/handlebars-version) was accepted by a dev
,automerge disabled, but the build was ok, so the PR got merged.
Later, more in depth tests show that the upgrade broke the app (I agree : a test is missing somewhere).
So the team reverted the commit to restore the old version.
And since, Renovate is opening PR for that upgrade. The team denies the new PR.. but the next run , renovate recreate the PR.

I came to this conclusion by reading the log & code : if i understood well Renovate says that it found a previous PR # 187 (the first one indeed) and as it is not rejected, recreate a new PR # 201.

I found that the [findPr](https://github.com/renovatebot/renovate/blob/main/lib/modules/platform/bitbucket/index.ts#L322) method for the BitBucket platform ask for the list of PRs, filter by branch, name & state and use the first, optional, item of this list.

I then modify the [getPrList](https://github.com/renovatebot/renovate/blob/main/lib/modules/platform/bitbucket/index.ts#L287) method to sort found PRs by pr number desc : last one first & run renovate manually.
So now the `findPr` found the PR 200 which was rejected and renovate stop poping new PR. (And other patch are processed as I have limited renovate to a single PR per repo)

I also found about the `reopen!` magic comment see https://github.com/renovatebot/renovate/pull/22984 and I wonder if a better solution exists with a `ignore!`  or other "magic" comment...
I still think that the last matching PR is more important than the first one... but I quickly look at the github platform impl and did not see a sort wo maybe I'm completely wrong.

Thanks !

### Logs (if relevant)

<details><summary>Logs</summary>

For visibility issue I have remove this common parts `"name":"renovate","hostname":"h","pid":9,"level":20,"logContext":"actx","repository":"acme/a-repo",`
```

{"msg":"Calculating hourly PRs remaining","time":"2024-11-15T16:23:07.067Z","v":0}
{"msg":"Initializing new PR cache at repository cache","time":"2024-11-15T16:23:07.067Z","v":0}
{"msg":"Syncing PR list","time":"2024-11-15T16:23:07.067Z","v":0}
{"msg":"http cache: saving https://api.bitbucket.org/2.0/repositories/acme/a-repo/pullrequests?state=OPEN&state=MERGED&state=DECLINED&state=SUPERSEDED&fields=values.id%2Cvalues.title%2Cvalues.state%2Cvalues.links.commits.href%2Cvalues.summary.raw%2Cvalues.source.branch.name%2Cvalues.destination.branch.name%2Cvalues.reviewers.display_name%2Cvalues.reviewers.uuid%2Cvalues.reviewers.nickname%2Cvalues.reviewers.account_status%2Cvalues.created_on%2Cvalues.updated_on&q=author.uuid+%3D+%22%7xxxxxxxx%7D%22&pagelen=50 (etag=\"d7c5bf4663d4e8789607410f99796eed\", lastModified=undefined)","time":"2024-11-15T16:23:15.600Z","v":0}
{"msg":"Fetched 28 PRs to sync with cache","time":"2024-11-15T16:23:15.600Z","v":0}
{"msg":"Total PRs cached: 28","time":"2024-11-15T16:23:15.676Z","v":0}
{"msg":"currentHourStart=2024-11-15T16:00:00.000+00:00","time":"2024-11-15T16:23:15.676Z","v":0}
{"msg":"PR hourly limit remaining: 19","time":"2024-11-15T16:23:15.676Z","v":0}
{"msg":"Calculating prConcurrentLimit (1)","time":"2024-11-15T16:23:15.677Z","v":0}
{"msg":"getBranchPr(renovate/handlebars-version)","time":"2024-11-15T16:23:15.677Z","v":0}
{"msg":"findPr(renovate/handlebars-version, undefined, open)","time":"2024-11-15T16:23:15.677Z","v":0}
--others pr listed-- 
{"msg":"0 PRs are currently open","time":"2024-11-15T16:23:15.679Z","v":0}
{"msg":"PR concurrent limit remaining: 1","time":"2024-11-15T16:23:15.679Z","v":0}
{"msg":"Calculated maximum PRs remaining this run: 1","time":"2024-11-15T16:23:15.679Z","v":0}
{"msg":"PullRequests limit = 1","time":"2024-11-15T16:23:15.679Z","v":0}
{"msg":"Calculating hourly PRs remaining","time":"2024-11-15T16:23:15.680Z","v":0}
{"msg":"currentHourStart=2024-11-15T16:00:00.000+00:00","time":"2024-11-15T16:23:15.680Z","v":0}
{"msg":"PR hourly limit remaining: 19","time":"2024-11-15T16:23:15.680Z","v":0}
{"msg":"Calculating branchConcurrentLimit (1)","time":"2024-11-15T16:23:15.680Z","v":0}
{"msg":"0 already existing branches found: ","time":"2024-11-15T16:23:15.681Z","v":0}
{"msg":"Branch concurrent limit remaining: 1","time":"2024-11-15T16:23:15.681Z","v":0}
{"msg":"Calculated maximum branches remaining this run: 1","time":"2024-11-15T16:23:15.681Z","v":0}
{"msg":"Branches limit = 1","time":"2024-11-15T16:23:15.681Z","v":0}
{"branch":"renovate/handlebars-version","msg":"syncBranchState()","time":"2024-11-15T16:23:15.681Z","v":0}
{"branch":"renovate/handlebars-version","msg":"syncBranchState(): Branch cache not found, creating minimal branchState","time":"2024-11-15T16:23:15.681Z","v":0}
{"branch":"renovate/handlebars-version","msg":"getBranchPr(renovate/handlebars-version)","time":"2024-11-15T16:23:15.682Z","v":0}
{"branch":"renovate/handlebars-version","msg":"findPr(renovate/handlebars-version, undefined, open)","time":"2024-11-15T16:23:15.682Z","v":0}
{"branch":"renovate/handlebars-version","msg":"branchExists=false","time":"2024-11-15T16:23:15.682Z","v":0}
{"branch":"renovate/handlebars-version","msg":"dependencyDashboardCheck=undefined","time":"2024-11-15T16:23:15.682Z","v":0}
{"branch":"renovate/handlebars-version","msg":"Check for closed PR because recreating closed PRs is disabled.","time":"2024-11-15T16:23:15.694Z","v":0}
{"branch":"renovate/handlebars-version","msg":"findPr(renovate/handlebars-version, Update dependency com.github.jknack:handlebars to v4.3.1, !open)","time":"2024-11-15T16:23:15.695Z","v":0}
{"branch":"renovate/handlebars-version","msg":"Found PR #187","time":"2024-11-15T16:23:15.695Z","v":0}
{"branch":"renovate/handlebars-version","msg":"Found closed PR with current title","time":"2024-11-15T16:23:15.695Z","v":0}
{"branch":"renovate/handlebars-version","msg":"Matching PR #187 was merged previously","time":"2024-11-15T16:23:16.239Z","v":0}
{"branch":"renovate/handlebars-version","msg":"Checking schedule(schedule=at any time, tz=Europe/Paris, now=2024-11-15T16:23:16.239Z)","time":"2024-11-15T16:23:16.239Z","v":0}
{"branch":"renovate/handlebars-version","msg":"No schedule defined","time":"2024-11-15T16:23:16.239Z","v":0}
{"branch":"renovate/handlebars-version","msg":"Branch needs creating","time":"2024-11-15T16:23:16.240Z","v":0}
{"branch":"renovate/handlebars-version","msg":"Using reuseExistingBranch: false","time":"2024-11-15T16:23:16.240Z","v":0}
{"branch":"renovate/handlebars-version","msg":"Setting current branch to master","time":"2024-11-15T16:23:16.240Z","v":0}
{"branch":"renovate/handlebars-version","branchName":"master","latestCommitDate":"2024-11-07T11:56:40+01:00","msg":"latest commit","time":"2024-11-15T16:23:16.264Z","v":0}
{"branch":"renovate/handlebars-version","msg":"manager.getUpdatedPackageFiles() reuseExistingBranch=false","time":"2024-11-15T16:23:16.278Z","v":0}
{"branch":"renovate/handlebars-version","msg":"Updating com.github.jknack:handlebars in pom.xml","time":"2024-11-15T16:23:16.286Z","v":0}
{"branch":"renovate/handlebars-version","msg":"updateArtifacts for updatedPackageFiles","time":"2024-11-15T16:23:16.286Z","v":0}
{"branch":"renovate/handlebars-version","msg":"Updated 1 package files","time":"2024-11-15T16:23:16.286Z","v":0}
{"branch":"renovate/handlebars-version","msg":"No updated lock files in branch","time":"2024-11-15T16:23:16.287Z","v":0}
{"branch":"renovate/handlebars-version","msg":"1 file(s) to commit","time":"2024-11-15T16:23:16.287Z","v":0}
{"branch":"renovate/handlebars-version","msg":"Preparing files for committing to branch renovate/handlebars-version","time":"2024-11-15T16:23:16.288Z","v":0}
{"branch":"renovate/handlebars-version","msg":"Setting git author name: Renovate Bot","time":"2024-11-15T16:23:16.288Z","v":0}
{"branch":"renovate/handlebars-version","msg":"Setting git author email: renovate@whitesourcesoftware.com","time":"2024-11-15T16:23:16.345Z","v":0}
{"branch":"renovate/handlebars-version","deletedFiles":[],"ignoredFiles":[],"result":{"author":null,"branch":"renovate/handlebars-version","commit":"1e52130f75bea353fed8dc836e165f55f9f31400","root":false,"summary":{"changes":1,"insertions":1,"deletions":1}},"msg":"git commit","time":"2024-11-15T16:23:16.474Z","v":0}
{"branch":"renovate/handlebars-version","msg":"Pushing refSpec renovate/handlebars-version:renovate/handlebars-version","time":"2024-11-15T16:23:16.480Z","v":0}
{"branch":"renovate/handlebars-version","result":{"pushed":[{"deleted":false,"tag":false,"branch":true,"new":true,"alreadyUpdated":false,"local":"refs/heads/renovate/handlebars-version","remote":"refs/heads/renovate/handlebars-version"}],"ref":{"local":"refs/remotes/origin/renovate/handlebars-version"},"remoteMessages":{"all":["Create pull request for renovate/handlebars-version:","https://bitbucket.org/acme/a-repo/pull-requests/new?source=renovate/handlebars-version&t=1"]}},"msg":"git push","time":"2024-11-15T16:23:27.568Z","v":0}
{"branch":"renovate/handlebars-version","msg":"Setting current branch to master","time":"2024-11-15T16:23:27.569Z","v":0}
{"branch":"renovate/handlebars-version","branchName":"master","latestCommitDate":"2024-11-07T11:56:40+01:00","msg":"latest commit","time":"2024-11-15T16:23:27.593Z","v":0}
{"name":"renovate","hostname":"h","pid":9,"level":30,"logContext":"actx","repository":"acme/a-repo","branch":"renovate/handlebars-version","commitSha":"1e52130f75bea353fed8dc836e165f55f9f31400","msg":"Branch created","time":"2024-11-15T16:23:27.607Z","v":0}
{"branch":"renovate/handlebars-version","msg":"Ensuring PR","time":"2024-11-15T16:23:27.607Z","v":0}
{"branch":"renovate/handlebars-version","msg":"There are 0 errors and 0 warnings","time":"2024-11-15T16:23:27.607Z","v":0}
{"branch":"renovate/handlebars-version","msg":"getBranchPr(renovate/handlebars-version)","time":"2024-11-15T16:23:27.608Z","v":0}
{"branch":"renovate/handlebars-version","msg":"findPr(renovate/handlebars-version, undefined, open)","time":"2024-11-15T16:23:27.608Z","v":0}
{"branch":"renovate/handlebars-version","msg":"Fetching changelog: https://github.com/jknack/handlebars.java (4.0.7 -> 4.3.1)","time":"2024-11-15T16:23:27.609Z","v":0}
{"branch":"renovate/handlebars-version","prTitle":"Update dependency com.github.jknack:handlebars to v4.3.1","msg":"Creating PR","time":"2024-11-15T16:23:27.686Z","v":0}
{"branch":"renovate/handlebars-version","title":"Update dependency com.github.jknack:handlebars to v4.3.1","base":"master","msg":"Creating PR","time":"2024-11-15T16:23:27.686Z","v":0}
{"branch":"renovate/handlebars-version","msg":"Adding PR #201 to the PR cache","time":"2024-11-15T16:23:37.644Z","v":0}
{"name":"renovate","hostname":"h","pid":9,"level":30,"logContext":"actx","repository":"acme/a-repo","branch":"renovate/handlebars-version","pr":201,"prTitle":"Update dependency com.github.jknack:handlebars to v4.3.1","msg":"PR created","time":"2024-11-15T16:23:37.645Z","v":0}

```

</details>

**Comment by RahulGautamSingh** [comment]
Closing a PR only stops renovate from creating PR for same version update. When another version will be available a new PR for same dep will be created.

Use `ignoreDeps` to completely stops update PRs.

**Comment by jgraglia** [comment]
Hi, thank you for your answer. 
> Closing a PR only stops renovate from creating PR for same version update

i'm well aware of that (for non immortal PRs) but my problem is that renovate still open the same PR (same branch name,  same PR title, marked as NON immortal) : `renovate/handlebars-version, Update dependency com.github.jknack:handlebars to v4.3.1`

because (my analysis) the first PR war accepted by error


renovate run :  
   (MARK 1) initial PR **187** is created with branch name & pr title 
   dev validate the pr & merge it : it was an error
   dev revert the merge commit to restore valid version
renovate run :  
   (MARK 2) another pr **192** is created with same branch name & pr title is created for the same dependency upgrade (   handlebars to v4.3.1)
   dev reject this one
 same for PR 193 & 194
 renovate run :  
   another pr 99*195** with same branch name & pr title is created for the same dependency upgrade (   handlebars to v4.3.1)
   another dev accept (unaware of the pb : not enought comm, not enough tests..)
   first dev revert  
renovate run :  
   another pr **198** with same branch name & pr title is created for the same dependency upgrade (   handlebars to v4.3.1)
And so on.. **199**, **200**, **201**...

heres is the PR of the renovate bot
![image](https://github.com/user-attachments/assets/030c6b6f-e2b3-4ac0-a177-9d8a54eea21f)

And thoses PR are NOT immortal (from my understanding)

the original PR that "break" the system (I think) PR187 : 
![image](https://github.com/user-attachments/assets/46d05ce1-d135-4022-8c7c-771878a4c7cc)

And the last one keep re created by renovte : PR201
![image](https://github.com/user-attachments/assets/b8290ea3-94d5-45e6-8c1f-9fdc7213c608)

**Comment by rarkins** [comment]
This may be because the findPr() for Bitbucket ~~Server~~ returns the _first_ matching closed PR (the merged one) and not the most recent one. We should check all platforms for this if so - I recall we fixed this in a different one previously (maybe GitHub).

**Comment by jgraglia** [comment]
Yes that is exactly what I was thinking about. And my successful "hack" was indeed to change findPr to find the newest one.

PS : my problem is with the **bitbucket cloud** platform, but as you said other platforms may have the problem

**Comment by jgraglia** [comment]
If you want I could share my workaround (after some polishing ;-) ) for the BitBucket cloud platform and share a PR ?

**Reply by RahulGautamSingh** [reply]
PR welcome

**Comment by rarkins** [comment]
What I think we should do is:
- Continue to store PRs in chronological order (oldest to newest)
- Use prList.findLast() to return the _last_ closed PR in cases like this

**Comment by jgraglia** [comment]
Hi, here is my PR https://github.com/renovatebot/renovate/pull/32672
@rarkins  I was not using `findLast` as I'm not sure that BitBucket will always return PR in ascending order : I prefer to manually sort (by pr number) to be more "indicative" of the intention, but of course I can change this.

I did not find a param to control the sort key in https://developer.atlassian.com/cloud/bitbucket/rest/api-group-pullrequests/#api-repositories-workspace-repo-slug-pullrequests-get
